jq255.md

```diff
@@ -0,0 +1,1425 @@
+
+  - A *challenge* is computed by hashing together the commitment `R`,
+    the public key `Q`, and the prepared message `M`, yielding a
+    value `c`.
+
+  - The *response* is the scalar `s = k + c*sk` (interpreting `c` as
+    an integer which is then converted to a scalar).
+
+The signature value itself consists of `c` and an encoding of `s`.
+
+The signature generation process is described as the `sign()` function
+below, using the private key (`sk`), public key (`Q`) and prepared
+message (`M`) as input. That function calls the `generate_nonce()`
+function to generate `k` in a safe and deterministic way; that process
+uses an extra optional parameter called `seed`. More on this function
+is detailed later on.
+
+```python
+def generate_nonce(sk, Q, M, seed):
+    sh = hashlib.blake2s()
+    sh.update(private_key_encode(sk))
+    sh.update(public_key_encode(Q))
+    sh.update(len(seed).to_bytes(8, byteorder='little'))
+    sh.update(seed)
+    sh.update(M)
+    bb = sh.digest()
+    return int_to_scalar(int.from_bytes(bb, byteorder='little'))
+
+def make_challenge(R, Q, M):
+    sh = hashlib.blake2s()
+    sh.update(group_encode(R))
+    sh.update(public_key_encode(Q))
+    sh.update(M)
+    return sh.digest()[0:16]  # 32-byte output is truncated to 16 bytes
+
+def sign(sk, Q, M, seed):
+    k = generate_nonce(sk, Q, M, seed)
+    R = point_mul(G, k)
+    c = make_challenge(R, Q, M)
+    ic = int.from_bytes(c, byteorder='little')
+    s = k + sk*int_to_scalar(ic)
+    return c + scalar_encode(s)    # Concatenation of c and encoded s
+```
+
+The `make_challenge()` function returns a 16-byte value. The signature
+value is the concatenation of `c` and the encoded scalar `s`, for
+a total length of 48 bytes.
+
+The per-signature nonce `k` is a scalar. The nonce may conceptually have
+value zero (though this is very improbable); thus, the commitment `R`
+may be the identity element. This is why `R` must be encoded with
+`group_encode()` and not `public_key_encode()` (public keys *cannot* be
+the identity element).
+
+The `generate_nonce()` function incarnates a process known as
+*derandomization*. Generally speaking, the nonce `k` should be generated
+randomly and uniformly among all possible scalar values. Any bias in
+this selection process may lead to private key reconstruction attacks.
+If a cryptographically secure random generator is available, then `k`
+may be obtained by simply getting 32 bytes out of that generator,
+interpreting them as a 256-bit integer, then reducing that integer
+modulo `r` (as in the private key generation process, this induces only
+negligible bias since `r` is close enough to a power of two). However,
+in practical implementations, a properly secure random generator is not
+necessarily available at signature generation time; the
+`generate_nonce()` function instead uses the private key, public key,
+message and optional extra seed to produce `k` in a safe way without
+requiring a random generator.
```

Comment:
You allow no attempt at sourcing randomness at all?

Reply:
There is a `seed` and you can choose the seed contents arbitrarily, including randomly, and that even yields benefits in terms of resistance against fault attacks. However, being able to provide a non-random seed allows reproducibility of test vectors.

Basically you _can_ provide a random seed; you do not _need_ to provide a random seed. There are situations where you prefer to inject randomness through the seed, and there are situations where you prefer a fully deterministic operation.

Reply:
@FiloSottile thoughts?